### PR TITLE
closure_js_library: fix: 'exports' now works in a data attr

### PR DESCRIPTION
**Pull request by robfig**
This adds a test that reproduces an issue I'm running into. I'm not sure what the fix is, so I figured I would raise it and see if you knew if this was expected or could provide any guidance on what's missing.

### CODE REVIEW

File: closure/testing/test/BUILD [hunk 1]
```diff
@@ -111,6 +111,46 @@ closure_js_test(
...
+closure_js_test(
+    name = "external_code_test",
+    srcs = ["external_code_test.js"],
+    html = "external_code_test.html",
+    entry_points = [
+        "rules_closure.ExternalCodeTest",
+    ],
+    data = [
+        # including external.js through a rollup target via exports should work.
+        ":all_external",
```
Comment (Yannic) [inline closure/testing/test/BUILD]: Does it work if you add `:external`?

I'm not very familiar with how the data attribute works. Does it consider `files`, `runfiles`, or both as input? If `runfiles` are used, then it should be easy to add. If it uses `files`, then it also shouldn't work if `:external` is a direct dependency, and this is working as intended.

https://github.com/bazelbuild/rules_closure/blob/308b05b2419edb5c8ee0471b67a40403df940149/closure/compiler/closure_js_library.bzl#L428

Comment (robfig) [reply closure/testing/test/BUILD]: Yes, it works if I reference the library directly. I'm afraid I don't know how it works either.

Comment (Yannic) [reply closure/testing/test/BUILD]: Okay, then, AFAICT, this is indeed a bug.
You should be able to fix it by adding `library.exports` to [this list](https://github.com/bazelbuild/rules_closure/blob/308b05b2419edb5c8ee0471b67a40403df940149/closure/compiler/closure_js_library.bzl#L432) :)

Comment (robfig) [reply closure/testing/test/BUILD]: Nailed it! That fixed the test. I've updated this pull request.

**Comment by robfig** [comment]
@dslomov @gkdn @laurentlb Please review, this is a small bug fix with unit test.

**Comment by robfig** [comment]
Good point, externs are not needed. Please take another look

File: closure/testing/test/BUILD [hunk 2]
```diff
@@ -111,6 +111,40 @@ closure_js_test(
     ],
 )
 
+# TEST: closure_js_library targets from exports should be provided when used in
+# a data attr.
+
+closure_js_library(
+    name = "external",
+    srcs = ["external.js"],
```
Comment (gkdn) [inline closure/testing/test/BUILD]: I think this part is little bit misleading. You are not trying to compile the code. That's why I recommended putting it on the data.

File: closure/testing/test/BUILD [hunk 3]
```diff
@@ -111,6 +111,40 @@ closure_js_test(
...
+)
+
+closure_js_library(
+    name = "all_external",
```
Comment (gkdn) [inline closure/testing/test/BUILD]: Let's put a more describing name; point is exporting right?

File: closure/testing/test/BUILD [hunk 4]
```diff
@@ -111,6 +111,40 @@ closure_js_test(
...
+    entry_points = [
+        "rules_closure.ExternalCodeTest",
+    ],
+    data = [
+        # including external.js through a rollup target via exports should work.
+        ":all_external",
+    ],
+    # NOTE: I'm not sure why this is necessary.
+    suppress = [
+        "reportUnknownTypes",
```
Comment (gkdn) [inline closure/testing/test/BUILD]: Is this still needed? What is unknown here?

File: closure/testing/test/BUILD (hunk 2 above)
Comment (robfig) [reply closure/testing/test/BUILD]: In my use case, I have a closure_js_library for each of my third party libraries, and they are compiled together using forgiving settings to create a vendor bundle. However, code that requires those libraries as externs needs to include them as data for their tests. I can change this test if you want, but then it wouldn't test my actual usage of this feature.

File: closure/testing/test/BUILD (hunk 3 above)
Comment (robfig) [reply closure/testing/test/BUILD]: `exported_external`?

File: closure/testing/test/BUILD (hunk 4 above)
Comment (robfig) [reply closure/testing/test/BUILD]: It's not still needed. Removing

File: closure/testing/test/BUILD (hunk 2 above)
Comment (gkdn) [reply closure/testing/test/BUILD]: > In my use case, I have a closure_js_library for each of my third party libraries, and they are compiled together using forgiving settings to create a vendor bundle. However, code that requires those libraries as externs needs to include them as data for their tests.

In that care you probably want a closure_js_binary and depend that as data in the test.

> I can change this test if you want, but then it wouldn't test my actual usage of this feature.

Your use case sounds non-trivial but the test should reflect basic underlying functionality/bug for maintenance purposes.

File: closure/testing/test/BUILD (hunk 3 above)
Comment (gkdn) [reply closure/testing/test/BUILD]: exports_data?

File: closure/testing/test/BUILD (hunk 2 above)
Comment (gkdn) [reply closure/testing/test/BUILD]: > In my use case, I have a closure_js_library for each of my third party libraries, and they are compiled together using forgiving settings to create a vendor bundle. However, code that requires those libraries as externs needs to include them as data for their tests.

In that care you probably want a closure_js_binary and depend that as data in the test.

> I can change this test if you want, but then it wouldn't test my actual usage of this feature.

Your use case sounds non-trivial but the test should reflect basic underlying functionality/bug for maintenance purposes.

Comment (gkdn) [reply closure/testing/test/BUILD]: > In my use case, I have a closure_js_library for each of my third party libraries, and they are compiled together using forgiving settings to create a vendor bundle. However, code that requires those libraries as externs needs to include them as data for their tests.

In that care you probably want a closure_js_binary and depend that as data in the test.

> I can change this test if you want, but then it wouldn't test my actual usage of this feature.

Your use case sounds non-trivial but the test should reflect basic underlying functionality/bug for maintenance purposes.

File: closure/testing/test/BUILD (hunk 3 above)
Comment (gkdn) [reply closure/testing/test/BUILD]: exports_data?

Comment (robfig) [reply closure/testing/test/BUILD]: updated

**Comment by gkdn** [comment]
Github is acting on me. Sorry for duplicates.

I noticed that the test is in in the wrong location. Need to be moved under closure/compiler/test/<scenario> and checking if something is on runfiles is easier to be done via sh_test:
[[ -e third_party/bazel_rules/rules_closure/closure/compiler/test/<scenario>/<data_file_name> ]]

File: closure/testing/test/BUILD (hunk 2 above)
Comment (robfig) [reply closure/testing/test/BUILD]: Yeah, those are still built with plovr, I suppose eventually we would use the binaries instead. I'll update it to just use data.

Comment (robfig) [reply closure/testing/test/BUILD]: Having a data-only closure_js_library target is not valid:

```
closure_js_library(
    name = "external",
    data = ["external.js"],
)
```

results in
```
Either 'srcs' or 'exports' must be specified
```

**Comment by robfig** [comment]
Sure thing. What do you want me to do with respect to using the `data` attribute instead of `srcs` for this test? Should I be using a different target type? 
https://github.com/bazelbuild/rules_closure/pull/391#discussion_r294370175

**Comment by gkdn** [comment]
> Sure thing. What do you want me to do with respect to using the `data` attribute instead of `srcs` for this test? Should I be using a different target type?
> [#391 (comment)](https://github.com/bazelbuild/rules_closure/pull/391#discussion_r294370175)

Let's make it general runfiles test and include but data and js srcs and verify both exists in the sh_test.

**Comment by robfig** [comment]
Updated, PTAL

**Comment by gkdn** [comment]
Thanks. Looks good.

**Comment by robfig** [comment]
Updated

**Pull request merged**